common/types/mapper/thrift/shared_test.go

```diff
@@ -3310,3 +3317,157 @@ func TestStickyWorkerUnavailableErrorConversion(t *testing.T) {
+			if len(orig.Value) == 0 {
+				return "empty data" // ignoring nil vs empty difference
+			}
+			return "filled data"
+		})
+	})
+
+	t.Run("can contain thrift data", func(t *testing.T) {
+		// pushing thrift or proto data through this type is very much expected,
+		// so this is both evidence that it's possible and an example for how to do so.
+		//
+		// note that there is an equivalent test in mapper/proto/shared_test.go.
+		// they are structurally the same, but encoding/decoding details vary a bit.
+
+		// some helpers because it's a bit verbose.
+		//
+		// note that these work for both types in this test: they can encode and decode *any* thrift data,
+		// you just have to make sure you pass the same type to both encode and decode via some other
+		// source of info (i.e. the ValueType field).
+		encode := func(thing thriftObject) []byte {
+			// get the intermediate format, ready for byte-encoding
+			wireValue, err := thing.ToWire()
+			require.NoError(t, err, "failed to produce intermediate wire.Value for encoding")
+			// write to binary
+			var writer bytes.Buffer
+			err = binary.Default.Encode(wireValue, &writer)
+			require.NoError(t, err, "failed to encode wire.Value")
+			// and return the bytes
+			result := writer.Bytes()
+			require.NotEmpty(t, result, "should have serialized some bytes")
+			return result
+		}
+		decode := func(data []byte, target thriftObject) {
+			// decode to the intermediate format, like encoding.
+			// the sample *shared.WorkflowExecution is a struct, so use wire.TStruct.
+			// if we used a map or int or something, it'd be a wire.TMap or wire.TI64, etc.
+			intermediate, err := binary.Default.Decode(bytes.NewReader(data), wire.TStruct)
+			require.NoError(t, err, "could not decode thrift bytes to intermediate wire.Value")
+			// and populate the target with the wire.Value contents
+			err = target.FromWire(intermediate)
+			require.NoError(t, err, "could not FromWire to the target type")
+		}
+
+		// --- create the original data, a thrift object, and encode it by hand
+		orig := &shared.WorkflowExecution{
+			WorkflowId: common.StringPtr(testdata.WorkflowID),
+			RunId:      common.StringPtr(testdata.RunID),
+		}
+		internalBytes := encode(orig)
+
+		// --- put that data into the custom Any type
+		// thrift has no registry like proto has, so there is no canonical name.
+		// no problem, just generate one.
+		var typeName = reflect.TypeOf(orig).PkgPath() + reflect.TypeOf(orig).Name()
+		anyVal := &types.Any{
+			ValueType: typeName,
+			Value:     internalBytes, // store thrift bytes in the Any
+		}
+
+		// --- convert the whole container to thrift (mimics making a call via yarpc)
+		thriftAny := FromAny(anyVal)      // we map to the rpc type
+		networkBytes := encode(thriftAny) // yarpc does this
+		// ^ this is what's sent over the network.
+
+		// the final data is not double-encoded, so this "encode -> wrap -> encode" process is reasonably efficient.
```

Comment:
why is it not double-encoded? I may have missed that part

Reply:
(also via chat but for public purposes...)

semantically: it doesn't matter, no behavior depends on it.

practically: because thrift (and I'm pretty sure proto) store binary blobs as `size, data...` so there's no need to e.g. escape `\0` as `\\0` or base64-encode it or other costly stuff.  it's just a memcpy.

my main goal here is to point out that anything you'd do with normal thrift is reasonable to do when wrapped by this `Any`-container because it doesn't meaningfully increase any of the runtime costs.  push gigabytes / million-field things through it, it's entirely fine and no different than using the non-wrapped type.

Reply:
I explained this more in the test comments now btw